### ISSUE TITLE
Abysmal performance in Robo Recall (as well as other Revive games IIRC)

### ISSUE THREAD

**Issue by wfunction** [LibreVR/Revive#525]
The performance is fine when nothing is going on but it tanks when enemies and/or lots of particle effects are on screen. Turning the graphics settings down doesn't seem to do much.

I've tried turning reprojection off to no effect. I also don't have any form of SS turned on (not in the settings file and not in the SteamVR [beta] settings). 

Other people on /r/vive are reporting the same problem, and some say the game runs fine on the Rift but horribly on the Vive despite using the same PC. 

My specs:

GTX 1070
i7 6700k 
16GB RAM

**Comment by hyperseven** [comment]
What other people on /r/vive? Cant say I have seen and threads about this.

**Comment by alivedoom** [comment]
Same here... as soon as I hit enemies i got massive fps drops... I just reinstalled the whole SteamVR and nvidia drivers and gonna try again.

My specs:
GTX 980Ti
i5 4670k
16GB RAM

**Comment by wfunction** [comment]
hyperseven: 

https://www.reddit.com/r/Vive/comments/5x9i1r/anyone_else_getting_abysmal_performance_in_robo/

alivedoom: 

Let me know if you stumble upon a fix! Are you using the beta branch?

**Comment by jmkraap** [comment]
I think this issue is related to this: https://github.com/LibreVR/Revive/issues/512

Hopefully Valve implements this in OpenVR soon or Revive may soon be unusable with newer titles.

**Comment by Armada651** [comment]
> Let me know if you stumble upon a fix! Are you using the beta branch?

Do *not* use the beta branch, it has a number of bugs that *will* break Revive. Stay on the stable branch.

@jmkraap It's very unlikely to be related, as the Unreal Engine does not usually suffer from CPU bottlenecks on their rendering backend.

**Comment by kvdv99** [comment]
using SteamVR but *not* the beta
I enabled Asynchr reprojection, but not interleaved
with 1080GTX
frame rate remains very high!
runs extremely smooth!

**Comment by alivedoom** [comment]
My performance got a tiny bit better by reinstalling the drivers though i remembered i had the front cam on and now its disabled so that gave that extra bit i think.

But NOT allowing interleaved reprojection actually did the job for me for now. Try it!!!!

**Comment by wfunction** [comment]
Turning off Interleaved did it for me too :) I didn't even realize I had it turned on as I thought I had turned it off awhile ago. 

It's playable now. Not 90fps across the board but 90fps most of the time which is good enough!

**Comment by 04nd01** [comment]
Something else I noticed that might be relevant to performance. I've had that dynamic resolution setting (don't remember the exact name) set to the default setting 0.7-1.0. And not once during the periods of bad performance did I notice the game actually lowering the res. When I set the max resolution to 0.7 I could immediately tell because everything looked horrible. Is it possible that this feature isn't working on revive?

**Comment by Armada651** [comment]
@04nd01 That is correct, Revive currently does not yet report a performance scale: https://github.com/LibreVR/Revive/blob/master/Revive/REV_CAPI.cpp#L823

**Comment by Werbel** [comment]
Im having FPS issues with Robo Recall.

It's weird because all world seems fine and hight fps, just enemies and weapons and destructibles seems to be low fps.

Everyone has it? I cant find anyone who described it like that...

**Comment by jaorg1234** [comment]
Do you have interleaved reprojection on? If it is on and the game goes down to 45 FPS, moving objects (including enemies) will have some kind of "ghosting" effect that may look like "low FPS" for you. Try turning that option off if it can maintain 90 FPS without it.

**Comment by Werbel** [comment]
> Well, thats why its weird - i was not looking into steamvr statistics but it is different than any other stuttering or fps drops or reprojection - ONLY enemies look lagged. I was not even sure if it was made on purpose by devs or its just me.

I'll try with reprojection off but im almost sure it will change nothing.

Im pretty well informed about how reprojection works and im really sensitive about it and can recognize even 1 frame drops or reprojection, but that was something i've seen first time.

**Comment by Armada651** [comment]
Since more and more Oculus Games are starting to make use of the Adaptive Performance Scale I'm going to implement it for the next version of Revive. I'll probably base it on the Adaptive Quality from The Lab Renderer:

https://github.com/ValveSoftware/the_lab_renderer/blob/ae64c48a8ccbe5406aba1e39b160d4f2f7156c2c/Assets/TheLabRenderer/Scripts/ValveCamera.cs#L631

**Comment by CarrotSurvivor** [comment]
Awesome man! How would that work though?

**Comment by Armada651** [comment]
@CarrotSurvivor Oculus has an API to get the quality scale, it's currently always set to 100% in Revive: https://github.com/LibreVR/Revive/blob/master/Revive/REV_CAPI.cpp#L817

So I can simply implement the adaptive scale algorithm from The Lab Renderer to calculate the correct scale based on the data provided by OpenVR.
